File: moderate/twenty_forty_eight.rb
def slide(a)
  last_num = last_num_id = last_zero = -1
  (0...a.length).each do |i|
    if a[i] == 0
      last_zero = i if last_zero == -1
      next
    elsif last_num == a[i]
      a[last_num_id] = 2 * last_num
      last_num = -1
      a[i] = 0
      last_zero = i if last_zero == -1
      next
    end
    last_num = a[i]
    if last_zero == -1
      last_num_id = i
      next
    end
    last_num_id = last_zero
    a[last_zero] = a[i]
    a[i] = 0
    last_zero += 1
  end
  a
end

File.open(ARGV[0]).each_line do |line|
  s = line.chomp.split('; ')
  n = s[1].to_i
  t = []
  s[2].split('|').each do |x|
    v = []
    x.split.each { |y| v << y.to_i }
    t << v
  end
  (0...n).each do |x|
    case s[0]
    when 'LEFT'
      t[x] = slide(t[x])
    when 'RIGHT'
      t[x] = slide(t[x].reverse).reverse
    when 'UP'
      v = []
      (0...n).each { |y| v << t[y][x] }
      v = slide(v)
      (0...n).each { |y| t[y][x] = v[y] }
    when 'DOWN'
      v = []
      (0...n).each { |y| v << t[n - y - 1][x] }
      v = slide(v)
      (0...n).each { |y| t[n - y - 1][x] = v[y] }
    end
  end
  (0...n).each { |x| t[x] = t[x].join(' ') }
  puts t.join('|')
end

File: moderate/twenty_forty_eight.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strings"
)

func slide(a []int) []int {
	lastNum, lastNumID, lastZero := -1, -1, -1
	for i := range a {
		if a[i] == 0 {
			if lastZero == -1 {
				lastZero = i
			}
		} else if lastNum == a[i] {
			a[lastNumID] = 2 * lastNum
			lastNum = -1
			a[i] = 0
			if lastZero == -1 {
				lastZero = i
			}
		} else {
			lastNum = a[i]
			if lastZero == -1 {
				lastNumID = i
			} else {
				lastNumID = lastZero
				a[lastZero] = a[i]
				a[i] = 0
				lastZero++
			}
		}
	}
	return a
}

func twentyFortyEight(n int, p, q string) string {
	t := strings.Split(q, "|")
	m := make([][]int, n)
	for i := 0; i < n; i++ {
		u := strings.Fields(t[i])
		m[i] = make([]int, n)
		for j := 0; j < n; j++ {
			fmt.Sscan(u[j], &m[i][j])
		}
	}
	l := make([]int, n)
	for i := 0; i < n; i++ {
		switch p {
		case "LEFT":
			copy(l, m[i])
			l = slide(l)
			copy(m[i], l)
		case "RIGHT":
			for j := 0; j < n; j++ {
				l[j] = m[i][n-1-j]
			}
			l = slide(l)
			for j := 0; j < n; j++ {
				m[i][n-1-j] = l[j]
			}
		case "UP":
			for j := 0; j < n; j++ {
				l[j] = m[j][i]
			}
			l = slide(l)
			for j := 0; j < n; j++ {
				m[j][i] = l[j]
			}
		case "DOWN":
			for j := 0; j < n; j++ {
				l[j] = m[n-1-j][i]
			}
			l = slide(l)
			for j := 0; j < n; j++ {
				m[n-1-j][i] = l[j]
			}
		}
	}
	r := make([]string, n)
	for i := 0; i < n; i++ {
		u := make([]string, n)
		for j := 0; j < n; j++ {
			u[j] = fmt.Sprint(m[i][j])
		}
		r[i] = strings.Join(u, " ")
	}
	return strings.Join(r, "|")
}

func main() {
	var n int
	data, err := os.Open(os.Args[1])
	if err != nil {
		log.Fatal(err)
	}
	defer data.Close()
	scanner := bufio.NewScanner(data)
	for scanner.Scan() {
		s := strings.Split(scanner.Text(), ";")
		fmt.Sscan(s[1], &n)
		fmt.Println(twentyFortyEight(n, s[0], s[2]))
	}
}
